### PR TITLE
Error backoff

### PR DESCRIPTION
**Pull request by ShanniLi**
r @jcorbin @Raynos @kriskowal @rf 
cc @junchaowu @prashantv @anson627 

The implementation is described in the design doc. Only linear backoff is implemented.
If this makes sense, we will need to implement it in python and go clients also.

### CODE REVIEW

File: node/error_backoff.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+var TypedError = require('error/typed');
+
+var BackoffError = TypedError({
```
Comment (anson627) [inline node/error_backoff.js]: move to errors.js?

File: node/error_backoff.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            serviceName: serviceName
+        });
+        return;
+    }
+
+    if (!self.shouldConsider(err)) {
+        return;
+    }
+
+    var edge = cn + '~~' + serviceName;
```
Comment (anson627) [inline node/error_backoff.js]: refactoring to a one line method, so that we duplicate the code.

File: node/request.js [hunk 3]
```diff
@@ -106,7 +106,7 @@ TChannelRequest.prototype.emitResponse = function emitResponse(res) {
     self.responseEvent.emit(self, res);
 };
 
-TChannelRequest.prototype.hookupStreamCallback = function hookupCallback(callback) {
+TChannelRequest.prototype.hookupStreamCallback = function hookupStreamCallback(callback) {
```
Comment (anson627) [inline node/request.js]: unrelated change?

File: node/error_backoff.js (hunk 1 above)
Comment (ShanniLi) [inline node/error_backoff.js]: It is a v2 specific error. Not sure errors.js is a good place. Ideally, it should be in v2/error_response.js, but then we will need to reference that file from the top level ...

File: node/request.js (hunk 3 above)
Comment (ShanniLi) [inline node/request.js]: it is in a separate commit. :)

File: node/channel.js [hunk 4]
```diff
@@ -255,6 +257,13 @@ function TChannel(options) {
         self.flushStats();
     }
 
+    if (!self.topChannel) {
+        self.errorBackoff = new ErrorBackoff({
```
Comment (kriskowal) [inline node/channel.js]: I recommend CongestionAvoidance (AIMD, controlled by errors), to be later complemented by FlowControl (controlled by latency). AIMD is "additive increase, multiplicative decrease". I see the multiplicative decrease part, but no mechanism for an edge to recover from failure. Be sure to trap successes as well as failures to complete the control loop.

File: node/error_backoff.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+ErrorBackoff.prototype.getBackoffError =
+function getBackoffError(cn, serviceName) {
+    var self = this;
+    if (!self.backoffRate) {
+        return null;
+    }
+
+    if (!cn || !serviceName) {
+        self.logger.warn('ErrorBackoff.getBackoffError called wiht invalid parameters', {
```
Comment (kriskowal) [inline node/error_backoff.js]: s/wiht/with/

File: node/error_backoff.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            cn: cn,
+            serviceName: serviceName
+        });
+        return null;
+    }
+
+    var edge = cn + '~~' + serviceName;
```
Comment (kriskowal) [inline node/error_backoff.js]: I know @raynos encourages this pattern as an expedient, but there is nothing at all, not even a doc, that prevents cn or sn from having this delimiter. This is why the circuit breaker tracks these in a two-level data structure. I encourage you to solve this problem by any means.

File: node/test/error_backoff.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+'use strict';
+
+var TimeMock = require('time-mock');
+var allocCluster = require('./lib/alloc-cluster.js');
+var timers = TimeMock(Date.now());
+var ErrorBackoff = require('../error_backoff.js');
+var series = require('run-series');
+
+function send(channel, endpoint, check, done) {
```
Comment (kriskowal) [inline node/test/error_backoff.js]: check arg is unused, probably also not useful since you can do the same in a callback thunk.

File: node/test/error_backoff.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        assert.ok(err, 'there should be an error');
+        assert.equal(err.type, 'tchannel.busy', 'error type as expected');
+    }
+    function check2(err, res) {
+        assert.ok(err, 'there should be a backoff error');
+        assert.equal(err.type, 'tchannel.busy', 'backoff error type as expected');
+        assert.equal(err.subType, 'tchannel.backoff.error', 'backoff error subType as expected');
+    }
+    function check3(err, res) {
+        assert.ifError(err, 'there should be no error');
```
Comment (kriskowal) [inline node/test/error_backoff.js]: I encourage everyone to use the pattern: `if (err) return assert.end(err);` instead of this err and continue pattern.

File: node/error_backoff.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+    if (!self.shouldConsider(err)) {
+        return;
+    }
+
+    var edge = cn + '~~' + serviceName;
+    if (!self.reqErrors[edge]) {
+        self.reqErrors[edge] = self.backoffRate;
+    } else {
+        self.reqErrors[edge] += self.backoffRate;
```
Comment (kriskowal) [inline node/error_backoff.js]: What does `reqErrors` represent? What are the units of `backoffRate`? This looks like linear accumulating backoff instead of multiplicative, but depends on what the unit are. Consider a more illuminating name, like `errorCountByEdge` (if count goes up by 1 for each error), or `delayByEdge` if the unit is an artificial delay (milliseconds) between requests.

(Please only abbreviate request to req in local variables.)

File: node/error_backoff.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        });
+        return null;
+    }
+
+    var edge = cn + '~~' + serviceName;
+    if (!self.reqErrors[edge] || self.reqErrors[edge] < 1) {
+        return null;
+    }
+
+    self.reqErrors[edge] -= 1;
```
Comment (kriskowal) [inline node/error_backoff.js]: I’m still reading and haven’t figured out the units, but this constant should probably have a name and be configurable, since backoffRate has a name and isn’t necessarily 1.

Is the idea that `getBackoffError` is depleting a penalty each time it’s called? `get` is usually reserved for pure accessors. Try `take` or `nextBackoffError` if it’s consuming something as a side-effect.

File: node/test/error_backoff.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        channel: channel,
+        backoffRate: 1
+    });
+    channel.logger.whitelist(
+        'warn',
+        'ErrorBackoff.handleError called wiht invalid parameters'
+    );
+    channel.logger.whitelist(
+        'warn',
+        'ErrorBackoff.getBackoffError called wiht invalid parameters'
```
Comment (kriskowal) [inline node/test/error_backoff.js]: s/wiht/with/

File: node/error_backoff.js (hunk 1 above)
Comment (kriskowal) [inline node/error_backoff.js]: We keep all errors in one module per package.

**Comment by kriskowal** [comment]
I am going to wait for a doc before commenting further. Intuition is that we either have per request back-off that forgets its state upon timeout, or per-edge back-off that has a corresponding back-on so we don’t lock a cn:sn pair into a degraded state.

File: docs/error-backoff.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
+# Error Backoff
+
+Error backoff is a mechanism that throttles outgoing TChannel traffic in order to improve the
+health of the Hyperbahn cluster and its hosted services. By definition, error backoff monitors
+the status of outgoing TChannel traffic. If it determines that 1) the Hyperbahn cluster/services
```
Comment (Raynos) [inline docs/error-backoff.md]: "the Hyperbahn cluster/services are busy, and 2) the busy state can be improved by throttling outgoing traffic"

Let's only throttle on Busy as first pass.

File: docs/error-backoff.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+are unhealthy, and 2) the unheathy state can be improved by throttling outgoing
+traffic, error backoff will effectively throttle outgoing traffic by returning error responses
+corresponding to the out requests.
+
+There are many reasons why error backoff can help. For example, when responding to `tchannel.busy`,
+it means that a service is stressed so that requests to the service should be reduced. As another
+example, when `tchannel.declined` is received, it may indicate the absence of any healthy host
```
Comment (Raynos) [inline docs/error-backoff.md]: tchannel.declined means "try me again somewhere else".

File: docs/error-backoff.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+for the service, so there is no reason to send more requests to the same service in the next short
+amount of time period.
+
+
+## Goals
+
+Error backoff works passively to defend Hyperbahn cluster and services from flooding requests and retries.
+When a service is unhealth or over stressed, there will be an increasing number of error responses.
```
Comment (Raynos) [inline docs/error-backoff.md]: The unhealthy case is completely different; that should be handled by the circuit breaker.

File: docs/error-backoff.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+Such an unhealthy state should be handled by reducing the requests to the service. On the other hand,
+it is very important that error backoff never affects healthy traffic.
+
+Goals:
+
+* Monitor the health state of an Hyperbahn edge, i.e., callerName ==> serviceName. Most effectively,
```
Comment (Raynos) [inline docs/error-backoff.md]: You are not the circuit breaker.

You should be monitoring the load on any given host:port; either by total or by service.

File: docs/error-backoff.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+  such information should be observed locally through call responses.
+* Throttle out requests if the current unheathy state can be relieved by reducing the requests.
```
Comment (Raynos) [inline docs/error-backoff.md]: You should throttle requests if you have determined that the host:port is busy, either by total or by service name.

File: docs/error-backoff.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+* The throttling should grow if the health state continues to decline.
```
Comment (Raynos) [inline docs/error-backoff.md]: The throttling should grow if we have a growth in busy frames.

File: docs/error-backoff.md [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+* The throttling should decline if the health state improves.
```
Comment (Raynos) [inline docs/error-backoff.md]: You are not health state; you are not circuit breaker. Backoff is purely about Busy frames.

File: docs/error-backoff.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+* Error backoff should converge. In other words, if the traffic stablizes in terms of failure/success
```
Comment (Raynos) [inline docs/error-backoff.md]: This is a great point; we definitely want this; if Backoff works correctly then the server should have at most 0/1 Busy's and the client should have a high rate of "ClientSide Busy"

File: docs/error-backoff.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+  ratios, the level of throttling should also be stable.
+* No healthy traffic should be affected at any time.
```
Comment (Raynos) [inline docs/error-backoff.md]: When a connection to hyperbahn has a single Busy service it should not affect any other services on said connection.

It should also not affect the caller as a whole.

File: docs/error-backoff.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+  on the Hyperbahn nodes, error backoff throttles the requests that can be forwarded through.
+
+
+## Design & Implemenation
+
+### Errors sensitive to backoff 
+Only issues that can be relieved by throttling should be considered for error backoff. Errors sensitive to
+backoff are also called eligible errors. There are three types of eligible errors handled by error backoff. 
+
+`tchannel.timeout` is received when a request is not completed within its timeout limit. The error indicates
```
Comment (Raynos) [inline docs/error-backoff.md]: timeout does not indicate that the service is slow; it indicates that a dependency somewhere is slow. The best solution is to try again or use speculative execution.

I dont think we need backoff on timeouts, we just need to smooth over them.

If a service is running at high CPU or at high event loop lagging and is leaking latency instead of spiking it; it needs to return busy frames.

File: docs/error-backoff.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+that the Hyerbahn cluster and/or the service is becoming slow.
+
+`tchannel.busy` is returned by Rate Limiting when one of the relaying nodes decides that the RPS limit for this
+request has been reached.
+
+`tchannel.declined` is returned when 1) there is no peer available for this service or 2) the request is declined
```
Comment (Raynos) [inline docs/error-backoff.md]: > Node declined request for reasons other than load, the request is safe to retry elsewhere if desired, do not change peer preferencing for future requests.

Declined means its safe to try elsewhere; circuit breaker should never have been returning declineds; we have an Unhealthy type for that.

File: docs/error-backoff.md [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+## Design & Implemenation
+
+### Errors sensitive to backoff 
+Only issues that can be relieved by throttling should be considered for error backoff. Errors sensitive to
+backoff are also called eligible errors. There are three types of eligible errors handled by error backoff. 
+
+`tchannel.timeout` is received when a request is not completed within its timeout limit. The error indicates
+that the Hyerbahn cluster and/or the service is becoming slow.
+
+`tchannel.busy` is returned by Rate Limiting when one of the relaying nodes decides that the RPS limit for this
```
Comment (Raynos) [inline docs/error-backoff.md]: Tchannel.busy is also returned by the edge server when it has high event loop lag / RAM / etc.

File: docs/error-backoff.md [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+request has been reached.
+
+`tchannel.declined` is returned when 1) there is no peer available for this service or 2) the request is declined
+by Circuit Breaking. 
+
+
+### Health monitor 
+Error backoff is specific to Hyperbahn edges, i.e., callerName ==> serviceName. Therefore, the measurement of
```
Comment (Raynos) [inline docs/error-backoff.md]: it is not specific to edges. It's either total or per serviceName. 

You may in the hyperbahn entry node case want to penalize incoming requests by callerName weighting.

File: docs/error-backoff.md [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+traffic health is associated with the edges.
+
+When an error is received, error backoff should take the error type and the frequency of the error into account.
+Only eligible errors should be considered towarding an unhealthy state. The frequency of the error is used to
```
Comment (Raynos) [inline docs/error-backoff.md]: Let's just stop using the word healthy; it's very confusing. Use the word "busy" or something.

File: node/v2/handler.js [hunk 26]
```diff
@@ -548,15 +549,17 @@ function sendCallRequestFrame(req, flags, args) {
         req.checksum.type, args
     );
 
-    if (!self.verifyCallRequestFrame(req, args)) {
-        return;
+    var result;
+    if (self.verifyCallRequestFrame(req, args)) {
```
Comment (Raynos) [inline node/v2/handler.js]: Why this change; let's avoid if possible.

File: docs/error-backoff.md (hunk 13 above)
Comment (ShanniLi) [inline docs/error-backoff.md]: if there is no peer available right now, does it actually mean try again later?

File: docs/error-backoff.md (hunk 14 above)
Comment (ShanniLi) [inline docs/error-backoff.md]: Yes, heathy state is confusing here. I will think about a better word.

The point I want express here is: the system is unheathy and error backoff my improve it. Error backoff doesn't really tracks the actually healthy state.

File: docs/error-backoff.md (hunk 15 above)
Comment (ShanniLi) [inline docs/error-backoff.md]: I don't think error backoff is congestion control. It should simply backoff on errors. It is rate limiter's job to monitor the load.

On the other hand, when enabled on Hyperbahn node, I think error backoff works as a mild "circuit breaker", i.e., reduce the traffic on a given edge.

File: docs/error-backoff.md (hunk 25 above)
Comment (kriskowal) [inline docs/error-backoff.md]: Busy and available. Backing off when busy, and recovering when available are the right words. Healthy and unhealthy are more general, covering all classes of error except deliberate cases like cancellation. What you are describing *is* flow control. Please consider treating this as a flow control problem, where busy errors are a signal of oversaturation, where a *multiplicative* back-off and successful delivery produces an *linear* recovery. Linear back-off won’t suffice.

Consider writing a *quick* simulation to verify the approach.

File: node/test/error_backoff.js (hunk 7 above)
Comment (ShanniLi) [inline node/test/error_backoff.js]: not sure if I understand. I will need to check the response at each series call, how can I do that without a check function?

File: node/channel.js (hunk 4 above)
Comment (ShanniLi) [inline node/channel.js]: it is actually error backoff only

File: docs/error-backoff.md (hunk 25 above)
Comment (ShanniLi) [inline docs/error-backoff.md]: Linear backoff is a good start for testing and evaluation. We can easily switch to exponential backoff when the framework of error backoff works.

File: docs/error-backoff.md [hunk 27]
```diff
@@ -0,0 +1,70 @@
+# Error Backoff
+
+Error backoff is a mechanism that throttles outgoing TChannel traffic in order to improve the
+health of the Hyperbahn cluster and its hosted services.
```
Comment (Raynos) [inline docs/error-backoff.md]: Improve availability

File: node/connection.js [hunk 28]
```diff
@@ -346,6 +346,8 @@ TChannelConnection.prototype.onCallError = function onCallError(err) {
 
         req.emitError(err);
     }
+
+    self.channel.errorBackoff.handleError(err, req.headers.cn, req.serviceName);
```
Comment (Raynos) [inline node/connection.js]: cleaner to just pass `req` in as a whole.

File: node/connection_base.js [hunk 29]
```diff
@@ -86,7 +86,13 @@ function connBaseRequest(options) {
     var req = self.buildOutRequest(options);
     self.ops.addOutReq(req);
     req.peer.invalidateScore();
+    req.errorEvent.on(onReqError);
+
     return req;
+
+    function onReqError(err) {
+        self.ops.popOutReq(req.id, req);
```
Comment (Raynos) [inline node/connection_base.js]: Why is this needed ? Do this in a seperate PR ?

File: node/errors.js [hunk 30]
```diff
@@ -72,6 +72,13 @@ module.exports.AsHeaderRequired = TypedError({
     frame: null
 });
 
+module.exports.BackoffError = TypedError({
+    type: 'tchannel.backoff.error',
+    message: 'Backoff error triggered due to previous requset failures {cn} ==> {serviceName}',
```
Comment (Raynos) [inline node/errors.js]: `request`

File: node/error_backoff.js [hunk 31]
```diff
@@ -0,0 +1,112 @@
...
+    if (!(this instanceof ErrorBackoff)) {
+        return new ErrorBackoff(options);
+    }
+    var self = this;
+
+    self.channel = options.channel;
+    assert(self.channel && !self.channel.topChannel, 'ErrorBackoff requires top channel');
+    self.backoffRate = options.backoffRate || 0;
+    assert(self.backoffRate >= 0, 'ErrorBackoff requires a nonnegative backoffRate');
+    self.reqErrors = {};
```
Comment (Raynos) [inline node/error_backoff.js]: `Object.create(null)`

File: node/channel.js [hunk 32]
```diff
@@ -255,6 +257,13 @@ function TChannel(options) {
...
+            channel: self,
+            backoffRate: self.options.backoffRate
```
Comment (Raynos) [inline node/channel.js]: Favor an enabled boolean over `backoffRate: 0`

File: node/error_backoff.js [hunk 33]
```diff
@@ -0,0 +1,112 @@
...
+
+ErrorBackoff.prototype.handleError =
+function handleError(err, cn, serviceName) {
+    var self = this;
+    if (!self.backoffRate) {
+        return;
+    }
+
+    if (!err || !cn || !serviceName) {
+        self.logger.warn('ErrorBackoff.handleError called with invalid parameters', {
```
Comment (Raynos) [inline node/error_backoff.js]: `self.logger.error()` please.

File: node/error_backoff.js [hunk 34]
```diff
@@ -0,0 +1,112 @@
...
+
+ErrorBackoff.prototype.nextBackoffError =
+function nextBackoffError(cn, serviceName) {
+    var self = this;
+    if (!self.backoffRate) {
+        return null;
+    }
+
+    if (!cn || !serviceName) {
+        self.logger.warn('ErrorBackoff.nextBackoffError called with invalid parameters', {
```
Comment (Raynos) [inline node/error_backoff.js]: `logger.error()`

File: node/out_request.js [hunk 35]
```diff
@@ -351,6 +351,12 @@ TChannelOutRequest.prototype.sendCallRequestFrame = function sendCallRequestFram
     var self = this;
     switch (self.state) {
         case States.Initial:
+            var errorBackoff = self.channel.errorBackoff || self.channel.topChannel.errorBackoff;
+            var err = errorBackoff.nextBackoffError(self.headers.cn, self.serviceName);
```
Comment (Raynos) [inline node/out_request.js]: favor `errorBackoff.nextBackoffError(self)`

File: node/error_backoff.js (hunk 10 above)
Comment (Raynos) [inline node/error_backoff.js]: +1 on named constant instead of magic number.

File: node/error_backoff.js [hunk 36]
```diff
@@ -0,0 +1,112 @@
...
+        return null;
+    }
+
+    var edge = getEdgeName(cn, serviceName);
+    if (!self.reqErrors[edge] || self.reqErrors[edge] < 1) {
+        return null;
+    }
+
+    self.reqErrors[edge] -= 1;
+    return errors.BackoffError({
```
Comment (Raynos) [inline node/error_backoff.js]: let's rename this to `shouldBackoff()` and construct the error in `out_request.js`

Then in out_request you can add way more useful context to this error.

**Comment by Raynos** [comment]
Let's get our good friend @jcorbin to chip in on this beast.

File: node/connection.js (hunk 28 above)
Comment (ShanniLi) [inline node/connection.js]: actually, this is clearer interface and easier for unit testing

File: node/connection_base.js (hunk 29 above)
Comment (ShanniLi) [inline node/connection_base.js]: without this fix, this PR will fail the test

File: node/out_request.js (hunk 35 above)
Comment (ShanniLi) [inline node/out_request.js]: the same reason, I want to keep the interface simple and easy for unit testing.

File: node/error_backoff.js [hunk 37]
```diff
@@ -0,0 +1,125 @@
...
+ErrorBackoff.prototype.type = 'tchannel.error-backoff';
+
+ErrorBackoff.prototype.handleError =
+function handleError(err, cn, serviceName) {
+    var self = this;
+    if (!self.enabled || !self.backoffRate) {
+        return;
+    }
+
+    if (!err || !cn || !serviceName) {
```
Comment (jcorbin) [inline node/error_backoff.js]: these look like programing errors to me, use assert

File: node/error_backoff.js [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+ErrorBackoff.prototype.disable =
+function disable() {
+    var self = this;
+    self.enabled = false;
+    self.reqErrors = Object.create(null);
+};
+
+ErrorBackoff.prototype.shouldConsider =
+function shouldConsider(err) {
+    if (err.type === 'tchannel.busy') {
```
Comment (jcorbin) [inline node/error_backoff.js]: Adding custom error classification logic here seems wrong to me.  Rather I think we want the exact same "is it (un)healthy" logic used by circuits, aka `errors.isUnhealthy(errors.classify(err))`

File: node/error_backoff.js [hunk 39]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+ErrorBackoff.prototype.shouldBackoff =
+function shouldBackoff(cn, serviceName) {
+    var self = this;
+    if (!self.enabled || !self.backoffRate) {
+        return false;
+    }
+
+    if (!cn || !serviceName) {
```
Comment (jcorbin) [inline node/error_backoff.js]: again, this seems a programming error to me, should be an assert?

File: node/request.js (hunk 3 above)
Comment (jcorbin) [inline node/request.js]: please break it out into separate pr, there's too much going on here as it is

File: node/connection_base.js (hunk 29 above)
Comment (jcorbin) [inline node/connection_base.js]: please pull this out into a separate pr first then, it is a separate bug

Comment (jcorbin) [inline node/connection_base.js]: also in that separate pr, please add a regression test, or have some existing test cover this prior bug so that we don't un-fix it ;-)

Comment (ShanniLi) [inline node/connection_base.js]: This bug can only be triggered with this PR's, because in this PR, we emit an error in TChannelOutRequest.prototype.sendCallRequestFrame at State.Initial

File: node/error_backoff.js (hunk 38 above)
Comment (ShanniLi) [inline node/error_backoff.js]: :+1: Good point!

**Comment by kriskowal** [comment]
We need to come to a consensus about the problem we are solving with this, and do more to vet whether this is the appropriate solution. The problem is congestion avoidance, and the solution should be a congestion avoidance algorithm. This implementation is shaped like a rate limiting / token bucket / leaky bucket solution, but it is not implemented in such terms and has at least one flaw (no cap on the penalty) that would be revealed by modeling the problem. We need to take this to the drawing board and get consensus before diving in.

Please,

1.) Extract the markdown document you’ve contributed and open a PR for just that. As written, it still needs work because it does not explain the proposed problem or solution. At least one of the non-goals is definitely the goal: congestion avoidance.
2.) Write a small simulation and verify the behavior of the proposed solution for the problems it is intended to solve. In this case, reducing traffic from ingress, through egress, and to a saturated service, regardless of where the bottleneck manifests, and recovering from a degraded state in a reasonable amount of time.
3.) Implement, test, deploy, repeat.

This particular problem is tricky, has a lot of prior-art, and deserves diligence.

**Comment by jcorbin** [comment]
:+1: @kriskowal , this feature requires more deliberate design, and especially modeling / simulation in a reboot.

**Comment by ShanniLi** [comment]
I was planning to implement a simple solution just like lb_pool or airlock. Not sure how the idea of congestion control gets in. Yes, it will have the effect of congestion control, but it is just one of its benefit.

**Comment by Raynos** [comment]
:+1:

This feature needs more thought; let's open a PR similar to @kriskowal PR for the hyperbahn worker selection logic that has a markdown and a simulation.

Let's get consensus on the algorithm and vet the simulation; once that's done we will move onto implementing; deploying to ingress; verification and more.

**Comment by ShanniLi** [comment]
Sounds good, I will create a separate PR with a MD to get a consensus about the design first.
